File: Ahorn/entities/BounceMoveBlock.jl
module BounceHelperBounceMoveBlock

using ..Ahorn, Maple

@mapdef Entity "BounceHelper/BounceMoveBlock" BounceMoveBlock(x::Integer, 
                                                              y::Integer, 
                                                              width::Integer=Maple.defaultBlockWidth, 
                                                              height::Integer=Maple.defaultBlockHeight,
                                                              direction::String="Right", 
                                                              speed::Number=60.0,
                                                              oneUse::Bool=false,
                                                              activationFlag::String="",
                                                              spritePath::String="objects/BounceHelper/bounceMoveBlock") 

directions = String[
    "Right",
	"UpRight",
	"Up",
	"UpLeft",
	"Left",
	"DownLeft",
	"Down",
	"DownRight",
	"Unknown"
]

const placements = Ahorn.PlacementDict(
    "Bounce Move Block (BounceHelper)" => Ahorn.EntityPlacement(
        BounceMoveBlock,
        "rectangle",
        Dict{String, Any}(
            "direction" => "Right"
        )
    )
)

Ahorn.editingOptions(entity::BounceMoveBlock) = Dict{String, Any}(
    "direction" => directions
)
Ahorn.minimumSize(entity::BounceMoveBlock) = 16, 16
Ahorn.resizable(entity::BounceMoveBlock) = true, true

Ahorn.selection(entity::BounceMoveBlock) = Ahorn.getEntityRectangle(entity)

midColor = (4, 3, 23) ./ 255
highlightColor = (59, 50, 101) ./ 255

const arrows = Dict{String, String}(
    "Right" => "/arrow00",
	"UpRight" => "/arrow01",
	"Up" => "/arrow02",
	"UpLeft" => "/arrow03",
	"Left" => "/arrow04",
	"DownLeft" => "/arrow05",
	"Down" => "/arrow06",
	"DownRight" => "/arrow07",
	"Unknown" => "/unknown"
)

function Ahorn.render(ctx::Ahorn.Cairo.CairoContext, entity::BounceMoveBlock, room::Maple.Room)
    x = Int(get(entity.data, "x", 0))
    y = Int(get(entity.data, "y", 0))


    width = Int(get(entity.data, "width", 32))
    height = Int(get(entity.data, "height", 32))

    tilesWidth = div(width, 8)
    tilesHeight = div(height, 8)

    direction = get(entity.data, "direction", "Right")
    spritePath = get(entity.data, "spritePath", "objects/BounceHelper/bounceMoveBlock")
    arrowSprite = Ahorn.getSprite(spritePath * arrows[direction], "Gameplay")

    frame = spritePath * "/base"

    Ahorn.drawRectangle(ctx, 2, 2, width - 4, height - 4, highlightColor, highlightColor)
    Ahorn.drawRectangle(ctx, 8, 8, width - 16, height - 16, midColor)

    for i in 2:tilesWidth - 1
        Ahorn.drawImage(ctx, frame, (i - 1) * 8, 0, 8, 0, 8, 8)
        Ahorn.drawImage(ctx, frame, (i - 1) * 8, height - 8, 8, 16, 8, 8)
    end

    for i in 2:tilesHeight - 1
        Ahorn.drawImage(ctx, frame, 0, (i - 1) * 8, 0, 8, 8, 8)
        Ahorn.drawImage(ctx, frame, width - 8, (i - 1) * 8, 16, 8, 8, 8)
    end

    Ahorn.drawImage(ctx, frame, 0, 0, 0, 0, 8, 8)
    Ahorn.drawImage(ctx, frame, width - 8, 0, 16, 0, 8, 8)
    Ahorn.drawImage(ctx, frame, 0, height - 8, 0, 16, 8, 8)
    Ahorn.drawImage(ctx, frame, width - 8, height - 8, 16, 16, 8, 8)
    Ahorn.drawRectangle(ctx, div(width - arrowSprite.width, 2) + 1, div(height - arrowSprite.height, 2) + 1, 8, 8, highlightColor, highlightColor)
    Ahorn.drawImage(ctx, arrowSprite, div(width - arrowSprite.width, 2), div(height - arrowSprite.height, 2))
end

end

File: Loenn/entities/bounceMoveBlock.lua
local drawableNinePatch = require("structs.drawable_nine_patch")
local drawableRectangle = require("structs.drawable_rectangle")
local drawableSprite = require("structs.drawable_sprite")

local BounceHelperBounceMoveBlock = {}

BounceHelperBounceMoveBlock.name = "BounceHelper/BounceMoveBlock"
BounceHelperBounceMoveBlock.depth = 8995
BounceHelperBounceMoveBlock.minimumSize = {16, 16}

local BounceHelperBounceMoveBlockDirections = {
    "Up", "UpRight", "Right", "DownRight", "Down", "DownLeft", "Left", "UpLeft", "Unknown"
}

BounceHelperBounceMoveBlock.fieldInformation = {
    direction = {
        options = BounceHelperBounceMoveBlockDirections,
        editable = false
    }
}

BounceHelperBounceMoveBlock.placements = {
    {
        name = "normal",
        data = {
            width = 16,
            height = 16,
            direction = "Right",
            speed = 60,
            oneUse = false,
            activationFlag = ""
        }
    },
    {
        name = "unknown",
        data = {
            width = 16,
            height = 16,
            direction = "Unknown",
            speed = 60,
            oneUse = false,
            activationFlag = ""
        }
    }
}

local ninePatchOptions = {
    mode = "border",
    borderMode = "repeat"
}

local buttonNinePatchOptions = {
    mode = "fill",
    border = 0
}

local midColor = {4 / 255, 3 / 255, 23 / 255}
local highlightColor = {59 / 255, 50 / 255, 101 / 255}
local buttonColor = {71 / 255, 64 / 255, 112 / 255}

local frameTexture = "/base"
local arrowTextures = {
    up = "/arrow02",
    upright = "/arrow01",
    right = "/arrow00",
    downright = "/arrow07",
    down = "/arrow06",
    downleft = "/arrow05",
    left = "/arrow04",
    upleft = "/arrow03",
    unknown = "/unknown"
}

function BounceHelperBounceMoveBlock.sprite(room, entity)
    local x, y = entity.x or 0, entity.y or 0
    local width, height = entity.width or 24, entity.height or 24

    local direction = string.lower(entity.direction or "up")
    local spritePath = entity.spritePath or "objects/BounceHelper/bounceMoveBlock"

    local blockTexture = spritePath .. frameTexture
    local arrowTexture = spritePath .. (arrowTextures[direction] or arrowTextures["up"])

    local ninePatch = drawableNinePatch.fromTexture(blockTexture, ninePatchOptions, x, y, width, height)

    local highlightRectangle = drawableRectangle.fromRectangle("fill", x + 2, y + 2, width - 4, height - 4, highlightColor)
    local midRectangle = drawableRectangle.fromRectangle("fill", x + 8, y + 8, width - 16, height - 16, midColor)

    local arrowSprite = drawableSprite.fromTexture(arrowTexture, entity)
    local arrowSpriteWidth, arrowSpriteHeight = arrowSprite.meta.width, arrowSprite.meta.height
    local arrowX, arrowY = x + math.floor((width - arrowSpriteWidth) / 2), y + math.floor((height - arrowSpriteHeight) / 2)
    local arrowRectangle = drawableRectangle.fromRectangle("fill", arrowX, arrowY, arrowSpriteWidth, arrowSpriteHeight, highlightColor)

    arrowSprite:addPosition(math.floor(width / 2), math.floor(height / 2))

    local sprites = {}

    table.insert(sprites, highlightRectangle:getDrawableSprite())
    table.insert(sprites, midRectangle:getDrawableSprite())

    for _, sprite in ipairs(ninePatch:getDrawableSprite()) do
        table.insert(sprites, sprite)
    end

    table.insert(sprites, arrowRectangle:getDrawableSprite())
    table.insert(sprites, arrowSprite)

    return sprites
end

return BounceHelperBounceMoveBlock